sylvia-derive/src/message.rs

```diff
@@ -595,6 +587,112 @@ impl<'a> MsgVariant<'a> {
             }
         }
     }
+
+    pub fn emit_querier_impl(&self) -> TokenStream {
+        let sylvia = crate_module();
+        let Self {
+            name,
+            fields,
+            return_type,
+            ..
+        } = self;
+
+        let parameters = fields.iter().map(MsgField::emit_method_field);
+        let fields_names = fields.iter().map(MsgField::name);
+        let variant_name = Ident::new(&name.to_string().to_case(Case::Snake), name.span());
+
+        #[cfg(not(tarpaulin_include))]
+        {
+            quote! {
+                fn #variant_name(&self, #(#parameters),*) -> Result< #return_type, #sylvia:: cw_std::StdError> {
+                    let query = QueryMsg:: #variant_name (#(#fields_names),*);
+                    self.querier.query_wasm_smart(self.contract, &query)
+                }
+            }
+        }
+    }
+
+    pub fn emit_querier_declaration(&self) -> TokenStream {
+        let sylvia = crate_module();
+        let Self {
+            name,
+            fields,
+            return_type,
+            ..
+        } = self;
+
+        let parameters = fields.iter().map(MsgField::emit_method_field);
+        let variant_name = Ident::new(&name.to_string().to_case(Case::Snake), name.span());
+
+        #[cfg(not(tarpaulin_include))]
+        {
+            quote! {
+                fn #variant_name(&self, #(#parameters),*) -> Result< #return_type, #sylvia:: cw_std::StdError>;
+            }
+        }
+    }
+}
+
+pub struct MsgVariants<'a>(Vec<MsgVariant<'a>>);
```

Comment:
I'm not sure if this is a good approach to generate Querier directly from MsgVariants.

Later (in https://github.com/CosmWasm/sylvia/issues/28#issuecomment-1549648466) when we will implement `From` on the `BoundQuerier` we will need to pass the `interfaces` to `emit_querier` or source and parse it there inside. It doesn't fit in `new` so this would be a double pass of source inside the `MsgVariants`.

We could introduce a new struct, f.e. `ParsedInput` or however we would call it that would hold all the attributes, messages, error type and so on. This would reduce the preparsing which we currently do in multiple places: `multitest`, `EnumMessage`, `ContractEnumMessage`.
This way we could use this struct (in some separate PR) in other Structs generating code and extract all the commons to be generated by the `ParsedInput` to avoid duplications and reduce maintanence cost.

Reply:
The thing is - whatever is generated able from msg variants itself should be generated from here so we do not expose fields. What you are mentioning is a matter of the responsibility split.

When you need to emit `From` implementation on `BoundQuerier`, you can do it in another, maybe a new type, as it does not need access to message variants (it needs only the modules where "forms" are implemented).

Same thing when you implement traits on queries - maybe you will need some additional data, but only for the `impl` block header, not for functions - you will be able to emit functions implementation from `MsgVariants`, which would be passed (by ref) to some additional structure, and then the this additional structure will generate inner implementation calling some `Msgvariants` emit-family function, and then it would add some context (like `impl` block itself).